# Lab-6 Template

Answer questions in the designated cells

H2O GBM tuning guide:  
https://github.com/h2oai/h2o-3/blob/master/h2o-docs/src/product/tutorials/gbm/gbmTuning.ipynb

## Preparation

Use dataset provided in the eLearning

In [1]:
import pandas as pd
pd.set_option('display.max_columns', 1500)

import warnings
warnings.filterwarnings('ignore')

#Extend cell width
from IPython.core.display import display, HTML
display(HTML("<style>.container { width:80% !important; }</style>"))

In [2]:
#Run once
#!pip install category_encoders
from category_encoders.target_encoder import TargetEncoder

### Load data

In [3]:
#train_df = pd.read_csv('./data/SBA_loans_small.zip')
X_train = pd.read_csv('SBA_loans_train.csv')
X_test  = pd.read_csv('SBA_loans_test.csv')

In [4]:
X_train.head(n=3)

City State    Zip                         Bank BankState   NAICS  \
0  Huntsville    AL  35811    BUSINESS LOAN CENTER, LLC        FL  621310   
1  SCOTTSDALE    AZ  85254  WELLS FARGO BANK NATL ASSOC        CA       0   
2      BANGOR    ME   4401          BANGOR SAVINGS BANK        ME  323110   

   Term  NoEmp  NewExist  CreateJob  RetainedJob  FranchiseCode  UrbanRural  \
0    73      1       2.0          2            1              0           1   
1    84      3       2.0          0            0              0           0   
2    84      9       1.0          0            0              1           1   

  RevLineCr LowDoc  DisbursementGross  BalanceGross    GrAppv  SBA_Appv  \
0         N      N            25000.0           0.0   25000.0   21250.0   
1         N      N            52000.0           0.0   52000.0   46800.0   
2         0      Y           150000.0           0.0  150000.0  127500.0   

   Defaulted  
0          1  
1          1  
2          0

In [5]:
print("Train shape:", X_train.shape)
print("Test shape:", X_test.shape)

Train shape: (337186, 20)
Test shape: (112396, 20)


In [6]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 337186 entries, 0 to 337185
Data columns (total 20 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   City               337177 non-null  object 
 1   State              337180 non-null  object 
 2   Zip                337186 non-null  int64  
 3   Bank               336587 non-null  object 
 4   BankState          336583 non-null  object 
 5   NAICS              337186 non-null  int64  
 6   Term               337186 non-null  int64  
 7   NoEmp              337186 non-null  int64  
 8   NewExist           337140 non-null  float64
 9   CreateJob          337186 non-null  int64  
 10  RetainedJob        337186 non-null  int64  
 11  FranchiseCode      337186 non-null  int64  
 12  UrbanRural         337186 non-null  int64  
 13  RevLineCr          335483 non-null  object 
 14  LowDoc             336198 non-null  object 
 15  DisbursementGross  337186 non-null  float64
 16  Ba

# Prepare Dataset

Replace missing values for all columns for both X_train and X_test.
Replace Na's with zero for numerical variables and with "Missing" for categorical

Encode Categorical variables using target encoder.

In [7]:
values_to_fill = {}
for col in X_train.drop(columns=['Defaulted']).columns:
    if X_train[col].dtype == 'object':
        values_to_fill[col] = "Missing"
    else:
        values_to_fill[col] = 0

X_train.fillna(value=values_to_fill,inplace=True)
X_test.fillna(value=values_to_fill, inplace=True)

cols_to_encode = []
cat_encoders = {}
for col in X_train.drop(columns=['Defaulted']).columns:
  if X_train[col].dtype == 'object':
    cols_to_encode.append(col)
    
#Add Zip to categoricals and convert to string in order to make encoder work
cols_to_encode.append("Zip")
X_train["Zip"] = X_train["Zip"].apply(str)
X_test["Zip"] = X_test["Zip"].apply(str)

te = TargetEncoder()
Y_tr = X_train["Defaulted"]
Y_tst = X_test["Defaulted"]
te.fit(X_train[cols_to_encode],Y_tr)

X_train_te = te.transform(X_train[cols_to_encode])
X_test_te = te.transform(X_test[cols_to_encode])

new_col_names = [col+"_te" for col in cols_to_encode]
X_train_te.columns = new_col_names
X_test_te.columns = new_col_names

X_tr = X_train_te.join(X_train)
X_tst = X_test_te.join(X_test)

X_tr.drop(columns=cols_to_encode+["Defaulted"],inplace=True, axis=1)
X_tst.drop(columns=cols_to_encode+["Defaulted"],inplace=True, axis=1)

In [8]:
X_tr.head(n=3)

City_te  State_te   Bank_te  BankState_te  RevLineCr_te  LowDoc_te  \
0  0.380952  0.167744  0.308181      0.158105      0.146342   0.186457   
1  0.191919  0.200634  0.138341      0.221678      0.146342   0.186457   
2  0.125984  0.096586  0.062500      0.076696      0.149252   0.090740   

     Zip_te   NAICS  Term  NoEmp  NewExist  CreateJob  RetainedJob  \
0  0.347826  621310    73      1       2.0          2            1   
1  0.228916       0    84      3       2.0          0            0   
2  0.088889  323110    84      9       1.0          0            0   

   FranchiseCode  UrbanRural  DisbursementGross  BalanceGross    GrAppv  \
0              0           1            25000.0           0.0   25000.0   
1              0           0            52000.0           0.0   52000.0   
2              1           1           150000.0           0.0  150000.0   

   SBA_Appv  
0   21250.0  
1   46800.0  
2  127500.0

In [9]:
X_tst.head(n=3)

City_te  State_te   Bank_te  BankState_te  RevLineCr_te  LowDoc_te  \
0  0.083335  0.224163  0.073593      0.222704      0.146342   0.090740   
1  0.128531  0.149706  0.138341      0.175423      0.146342   0.186457   
2  0.170213  0.224163  0.000000      0.222704      0.146342   0.186457   

     Zip_te  NAICS  Term  NoEmp  NewExist  CreateJob  RetainedJob  \
0  0.090913      0    84      1       2.0          0            0   
1  0.170732      0    84      7       1.0          0            0   
2  0.229167      0   240     19       1.0         15            0   

   FranchiseCode  UrbanRural  DisbursementGross  BalanceGross    GrAppv  \
0              1           0            42000.0           0.0   42000.0   
1              1           0            15000.0           0.0   15000.0   
2              1           0           497000.0           0.0  497000.0   

   SBA_Appv  
0   33600.0  
1   13500.0  
2  497000.0

In [10]:
print("Train shape:", X_tr.shape)
print("Test shape:", X_tst.shape)

Train shape: (337186, 19)
Test shape: (112396, 19)


## Datasets for all questions

For all questions, use X_tr and X_tst (after categorical variables encoding).

## Question 1 - 2 points

Train sklearn `GradientBoostingClassifier` with default parameters and `random_state=0`.
Display:
- AUC on Testing data
- Accuracy on Testing data
- Number of trees for the trained classifier

In [11]:
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.metrics import roc_auc_score,accuracy_score,f1_score,precision_score,recall_score
clf = GradientBoostingClassifier(random_state = 0).fit(X_tr, Y_tr)


In [12]:
y_true = list(Y_tst)
y_tst_pred = clf.predict(X_tst)

In [13]:
result = pd.DataFrame({"Gradient Boosting with default parameters":[accuracy_score(y_true,y_tst_pred),
                                                                  roc_auc_score(y_true,y_tst_pred),
                                                                  f1_score(y_true,y_tst_pred,average='weighted'),
                                                                  precision_score(y_true,y_tst_pred),
                                                                  recall_score(y_true,y_tst_pred)]},index =['Accuracy Score','AUC score','F1 score',
                                                                                                          'precision score','Recall score'])

result

Gradient Boosting with default parameters
Accuracy Score                                    0.919036
AUC score                                         0.819500
F1 score                                          0.915259
precision score                                   0.838860
Recall score                                      0.666210

The AUC on testing data is 0.819500. 
The Accuracy on testing data is 0.919036

In [14]:
print("The number of trees for the trained classifier:",clf.n_estimators)         

The number of trees for the trained classifier: 100


## Question 2 - 3 points

Train sklearn `GradientBoostingClassifier` with following parameters:
```
n_estimators=1000, 
learning_rate=0.1,
subsample=0.8,
max_features=0.8,
n_iter_no_change=5,
max_depth=3, 
random_state=0
```

Display:
- AUC on Testing data
- Accuracy on Testing data
- Number of trees for the trained classifier

In [15]:
clf_new = GradientBoostingClassifier(n_estimators = 1000, learning_rate = 0.1,
                                     subsample = 0.8, max_features = 0.8,
                                     n_iter_no_change = 5, max_depth = 3,
                                     random_state = 0).fit(X_tr, Y_tr)

In [16]:
y_tst_predict = clf_new.predict(X_tst)

In [17]:
result = pd.DataFrame({"Gradient Boosting with default parameters":[accuracy_score(y_true,y_tst_predict),
                                                                  roc_auc_score(y_true,y_tst_predict),
                                                                  f1_score(y_true,y_tst_predict,average='weighted'),
                                                                  precision_score(y_true,y_tst_predict),
                                                                  recall_score(y_true,y_tst_predict)]},index =['Accuracy Score','AUC score','F1 score',
                                                                                                          'precision score','Recall score'])

result

Gradient Boosting with default parameters
Accuracy Score                                    0.933005
AUC score                                         0.855718
F1 score                                          0.930903
precision score                                   0.861142
Recall score                                      0.736693

The AUC on testing data is 0.855718. The Accuracy on testing data is 0.933005

In [18]:
print("The number of trees for the trained classifier:",clf_new.n_estimators)         

The number of trees for the trained classifier: 1000


## Question 3 - 10 points

Use Grid search to train at least 16 `GradientBoostingClassifier`. Tune following parameters, set `random_state=0`:
```
n_estimators, 
learning_rate,
subsample,
max_features,
n_iter_no_change,
max_depth, 
```
Your grid search should be performed using CV=3.
To speed up training process, set `n_jobs` to number of available cores on your machine. For example, if you have 8 CPU/Core machine, set `n_jobs=4` or `n_jobs=6`

For the best model (based on AUC) display:
- Model parameters
- AUC on Testing data
- Accuracy on Testing data
- Number of trees for the trained classifier

**Important**: It will take long time to train models, you will be training at least 16x3=48 models. Test that your Grid search works with small subset of the data first.

Once you have trained the Grid object, save it to the disk so that you can retrieve  it without going through re-train step. You can find example on how to save model in the Project 1 template.

For the best model, display:
- AUC on Testing data
- Accuracy on Testing data
- Number of trees for the trained classifier

**Optional** questions (if you plan to practice data science, you should make every effort to answer them): 
- Why do you think number of trees for the best model is less than `n_estimators`?
- Think about all parameters you are tuning using grid search.
- How are those parameters help to reduce overfit? 

In [19]:
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import make_scorer

In [20]:
gbc = GradientBoostingClassifier(random_state =0)
#gbc_default = GradientBoostingClassifier().fit(X_tr, Y_tr)

In [25]:
param_gbc = {
    'n_estimators':[200,400],
    'learning_rate':[0.2,0.3],
    'subsample':[0.5,1.0],
    'max_features':['sqrt'],
    'n_iter_no_change':[3,4],
    'max_depth':[4,5]
}



Grid_gbc = GridSearchCV(estimator=gbc, param_grid = param_gbc, scoring='roc_auc',cv = 3, n_jobs = 2)



max_depth --> To control over-fitting as higher depth will allow model to learn relations very specific to a particular sample.
max_features --> square root of the total number of features works great.
learning_rate --> Lower values make the model robust to the specific characteristics of tree and allows it to generalize well.
n_estimators --> Though GBM is fairly robust at higher number of trees but it can still overfit at a point. so,it should be tuned using CV for a particular learning rate.
subsample --> Values slightly less than 1 make the model robust by reducing the variance.


In [26]:
grid_fit = Grid_gbc.fit(X_tr, Y_tr)

In [27]:
y_test_predicted = grid_fit.predict(X_tst)

In [28]:
result_final = pd.DataFrame({"Gradient Boosting with default parameters":[accuracy_score(y_true,y_test_predicted),
                                                                  roc_auc_score(y_true,y_test_predicted),
                                                                  f1_score(y_true,y_test_predicted,average='weighted'),
                                                                  precision_score(y_true,y_test_predicted),
                                                                  recall_score(y_true,y_test_predicted)]},index =['Accuracy Score','AUC score','F1 score',
                                                                                                          'precision score','Recall score'])

result_final

Gradient Boosting with default parameters
Accuracy Score                                    0.936012
AUC score                                         0.861277
F1 score                                          0.934032
precision score                                   0.870376
Recall score                                      0.746182

In [29]:
grid_fit.best_params_

{'learning_rate': 0.2,
 'max_depth': 5,
 'max_features': 'sqrt',
 'n_estimators': 400,
 'n_iter_no_change': 4,
 'subsample': 1.0}

In [30]:
grid_fit.best_params_.get(accuracy_score)

In [31]:
grid_fit.best_score_

0.9802177584595412

## Question 4 - 5 points

Train Stacked Ensemble model by utilizing `from sklearn.ensemble import StackingClassifier`.

Good guide on how to train Stacked ensemble model can be found here: https://machinelearningmastery.com/stacking-ensemble-machine-learning-with-python/


Stacked Ensemble is a technique to build Meta Learner (model). The Meta learner uses out-of-fold predictions of level-0 models (at least 2) to train level-1 model(meta learner). Meta learner is trained on the out-of-fold predictions done by the level-0 model in order to avoid overfitting.  

For example, if level-0 model was trained using cv=3 it means there are 3 level-0 sub-models each trained on the 2/3 of the data. Therefore the hold-out 1/3 part of the training data was not used by the model for training, and predictions on the out of fold parts of the dataset can be used by Meta Learner. 

Choose two models for the level-0 models:
- Best model from the Question 3
- Worst model from the question 3

You would only need model parameters, since sklearn `StackingClassifier` will retrain the the models.

Train `StackingClassifier` and produce:
- AUC on Testing data
- Accuracy on Testing data


*Hint*: to find best/worst model parameters for your `grid_search` object you can use below code. You would need to change `grid_search` to the variable name that holds your `GridSearchCV` object. 
```
import numpy as np

best_model_idx = np.argmax(grid_search.cv_results_['mean_test_score'])
worst_model_idx = np.argmin(grid_search.cv_results_['mean_test_score']) 
print("Index of worst model:",worst_model_idx)
print("Index of best model:",best_model_idx)

print("Best model params:")
print(grid_search.cv_results_['params'][best_model_idx])
print("")
print("Worst model params:")
print(grid_search.cv_results_['params'][worst_model_idx])
```

In [38]:
from sklearn.ensemble import StackingClassifier
from sklearn.linear_model import LogisticRegression
import numpy as np

In [41]:
best_model_idx = np.argmax(grid_fit.cv_results_['mean_test_score'])
worst_model_idx = np.argmin(grid_fit.cv_results_['mean_test_score']) 
print("Index of worst model:",worst_model_idx)
print("Index of best model:",best_model_idx)

print("Best model params:")
print(grid_fit.cv_results_['params'][best_model_idx])
print("")
print("Worst model params:")
print(grid_fit.cv_results_['params'][worst_model_idx])

Index of worst model: 16
Index of best model: 15
Best model params:
{'learning_rate': 0.2, 'max_depth': 5, 'max_features': 'sqrt', 'n_estimators': 400, 'n_iter_no_change': 4, 'subsample': 1.0}

Worst model params:
{'learning_rate': 0.3, 'max_depth': 4, 'max_features': 'sqrt', 'n_estimators': 200, 'n_iter_no_change': 3, 'subsample': 0.5}


In [42]:
level0 = list()
level0.append(('bm',  GradientBoostingClassifier(learning_rate= 0.2, max_depth= 5, max_features= 'sqrt', n_estimators= 400, n_iter_no_change= 4, subsample= 1.0)))
level0.append(('wm', GradientBoostingClassifier(learning_rate= 0.3, max_depth= 4, max_features= 'sqrt', n_estimators= 200, n_iter_no_change= 3, subsample= 0.5)))

# define meta learner model
level1 = LogisticRegression()
# define the stacking ensemble
model = StackingClassifier(estimators=level0, final_estimator=level1, cv=3)
# fit the model on all available data
model.fit(X_tr, Y_tr)

StackingClassifier(cv=3,
                   estimators=[('bm',
                                GradientBoostingClassifier(learning_rate=0.2,
                                                           max_depth=5,
                                                           max_features='sqrt',
                                                           n_estimators=400,
                                                           n_iter_no_change=4)),
                               ('wm',
                                GradientBoostingClassifier(learning_rate=0.3,
                                                           max_depth=4,
                                                           max_features='sqrt',
                                                           n_estimators=200,
                                                           n_iter_no_change=3,
                                                           subsample=0.5))],
                   final_estimator=LogisticRegre

In [43]:
pred_stack = model.predict(X_tst) # predictions on test data using stacked model

In [44]:
result = pd.DataFrame({"Gradient Boosting with default parameters":[accuracy_score(y_true,pred_stack),
                                                                  roc_auc_score(y_true,pred_stack),
                                                                  f1_score(y_true,pred_stack,average='weighted'),
                                                                  precision_score(y_true,pred_stack),
                                                                  recall_score(y_true,pred_stack)]},index =['Accuracy Score','AUC score','F1 score',
                                                                                                          'precision score','Recall score'])

result

Gradient Boosting with default parameters
Accuracy Score                                    0.935336
AUC score                                         0.859668
F1 score                                          0.933300
precision score                                   0.869088
Recall score                                      0.743137

The AUC on testing data is 0.859668. The Accuracy on testing data is 0.935336